webapp/migrations/Version20240601180624.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+declare(strict_types=1);
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20240601180624 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'Move data_source to shadow_mode';
+    }
+
+    public function up(Schema $schema): void
+    {
```

Comment:
No other migration required?

Reply:
I think the easy fix is to tell people the query they have to run to check if they would get conflicts? And if we have that query we can already run it for them (hopefully in 1 transaction) and fail the migration if we encounter that situation?

Reply:
How would you propose this? I don't think we can do that in a query but only in code, i.e. load all entities and check if it would give an issue. Then if it does, what then? I don't think we can output log/stdout messages in a migration.

And the event feed still remains.

Reply:
Can't we do a "Select COUNT(externalid) from table t leftjoin t2 on t.id=t.externalid where t.externalid=NULL" (did not check the query) which people manually run and see if they have a problem?

When that's possible we can also check how to do this in a migration (but this should be a cornercase IMO which people can manually fix if they know for which entities to do it). If it happens we can just hardfail and stop migrating.

For the eventfeed we would need to update a lot of events right? Let's handle that case after this "simple" migration.

Reply:
That would maybe work: run that check in the migration, make it fail if there is something wrong. People fix it, rerun the migration.

Reply:
I cooked up a fancy migration 👀

Reply:
I would not fix the event feed indeed, since I see no easy way to do this. Not many people use it, even less who also upgrade DOMjudge. We can bump the version number indeed to indicate a breaking change.